test/rules.jl

```diff
@@ -0,0 +1,77 @@
+using Enzyme
+using Enzyme: EnzymeRules
+using Test
+
+import EnzymeRules: forward
+
+@testset "Custom Forward Rules" begin
+
+    rule_f(x) = x^2
+
+    function rule_fip(x)
+       x[1] *= x[1]
+       return nothing
+    end
+
+    function forward(::Type{Tuple{typeof(rule_f), Float64}}, rt, args)
```

Comment:
This seems a bit awkward. It really doesn't jive with multiple dispatch that we have to do subtype queries inside the rule.

```
function forward(::Const{typeof(rule_f)}, RT::Type{<:DuplicatedNoNeed}, x::Duplicated)
    return 10+2*x.val*x.dval
end

function forward(::Const{typeof(rule_f)}, RT::Type{<:BatchDuplicatedNoNeed}, x::BatchDuplicated{T, N}) where {T, N}
    return NTuple{N, T}(1000+2*x.val*dv for dv in x.dval)
end

function forward(::Const{typeof(rule_f)}, RT::Type{Duplicated}, x::Duplicated)
    return Duplicated(func.val(x.val), 100+2*x.val*x.dval)
end

function forward(::Const{typeof(rule_f)}, RT::Type{DuplicatedNoNeed}, x::BatchDuplicated{T, N}) where {T,N}
    return BatchDuplicated(func.val(x.val), NTuple{N, T}(10000+2*x.val*dv for dv in x.dval))
end
```

Would seem more julian to me.